kubeadm: do not use --admission-control for the API server

neolit123 opened the pull request:
**What this PR does / why we need it**:
The API server argument --admission-control is deprecated.
Use the following arguments instead:
  --enable-admission-plugins=NodeRestriction
  --disable-admission-plugins=PersistentVolumeLabel

Add comment that PersistentVolumeLabel should be removed at some
point in 1.11.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Updates kubernetes/kubeadm#840

**Special notes for your reviewer**:
NONE

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
kubeadm: when starting the API server use the arguments --enable-admission-plugins and --disable-admission-plugins instead of the deprecated --admission-control.
```

@luxas
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/area kubeadm

neolit123 committed:
kubeadm: do not use --admission-control for the API server

The API server argument --admission-control is deprecated.
Use the following arguments instead:
  --enable-admission-plugins=NodeRestriction
  --disable-admission-plugins=PersistentVolumeLabel

Add comment that PersistentVolumeLabel should be removed at some
point in 1.11.

luxas commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -143,7 +143,8 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration) []string {
 	defaultArguments := map[string]string{
 		"advertise-address":               cfg.API.AdvertiseAddress,
 		"insecure-port":                   "0",
-		"admission-control":               defaultAdmissionControl,
```
please remove the unused variable

neolit123 replied:
will do.

timothysc commented on cmd/kubeadm/app/phases/controlplane/manifests_test.go:
```diff
@@ -154,7 +154,8 @@ func TestGetAPIServerCommand(t *testing.T) {
 				"--insecure-port=0",
-				"--admission-control=NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota",
+				"--enable-admission-plugin=NodeRestriction",
+				"--disable-admission-plugin=PersistentVolumeLabel",
```
If this is deprecated I don't know why we need to explicitly specify.

neolit123 replied:
@luxas can comment further i guess, i was just following the guide here:
https://github.com/kubernetes/kubeadm/issues/840

luxas replied:
It's still enabled by default which is a bug. Until that bug is fixed, let's leave it out here.

timothysc replied:
I'm fine with defaulting off, but lets cross-link an issue.

neolit123 replied:
should i cross link to this?:
https://github.com/kubernetes/kubeadm/issues/840

timothysc replied:
@neolit123 I'm thinking about the api-server deprecation default issue, should be in the main repo somewhere.

neolit123 replied:
i can't seem to find it. only found the PR that deprecated `admission-control`: https://github.com/kubernetes/kubernetes/pull/58123

luxas replied:
https://github.com/kubernetes/kubernetes/pull/64326 and https://github.com/kubernetes/kubernetes/issues/52617 and https://github.com/kubernetes/kubernetes/pull/52618

luxas reviewed (approved):
Thanks @neolit123! Can you use kubeadm with this patch to start up both a v1.10 cluster and a v1.11-beta.0 API server and post the final admission chain that was loaded (it's printed in the API server logs in the beginning), so we have a double-check that everything we want to be running actually is?
/approve

neolit123 commented:
> Thanks @neolit123! Can you use kubeadm with this patch to start up both a v1.10 cluster and a v1.11-beta.0 API server and post the final admission chain that was loaded (it's printed in the API server logs in the beginning), so we have a double-check that everything we want to be running actually is?
/approve

ha!! as long as my control plane even starts. my setup is beyond broken ATM and it's not really kubeadm's fault. i can try...

stealthybox commented:
@neolit123 if your personal lab is trashed, this should serve the need:
https://github.com/kubernetes/kubeadm/tree/master/vagrant

just change the package versions in the `Vagrantfile` to 1.10.2 instead of 1.9.7 and then follow the guide for comparing different builds with the already installed kubeadm binary

neolit123 commented:
yeah @stealthybox . we spoke about it at KubeCon.
i didn't had the time to try it, but i definitely need to find the time...

neolit123 commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -143,7 +141,9 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration) []string {
 		"advertise-address":               cfg.API.AdvertiseAddress,
 		"insecure-port":                   "0",
-		"admission-control":               defaultAdmissionControl,
+		// https://github.com/kubernetes/kubeadm/issues/840
```
^ added link to the issue here.

neolit123 commented:
/test pull-kubernetes-kubemark-e2e-gce-big 
/test pull-kubernetes-node-e2e

neolit123 commented:
/test pull-kubernetes-node-e2e
/test pull-kubernetes-kubemark-e2e-gce-big

luxas commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -141,9 +139,11 @@ func createStaticPodFiles(manifestDir string, cfg *kubeadmapi.MasterConfiguratio
-		"admission-control":               defaultAdmissionControl,
+		"advertise-address": cfg.API.AdvertiseAddress,
+		"insecure-port":     "0",
+		// https://github.com/kubernetes/kubernetes/pull/58123
```
I'm not sure we should have this here...

luxas commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -141,9 +139,11 @@ func createStaticPodFiles(manifestDir string, cfg *kubeadmapi.MasterConfiguratio
+		"insecure-port":     "0",
+		// https://github.com/kubernetes/kubernetes/pull/58123
+		"enable-admission-plugin":         "NodeRestriction",
+		"disable-admission-plugin":        "PersistentVolumeLabel",
```
Add a TODO to remove this in kubeadm v1.11, as it's automatically disabled in v1.11, and reference https://github.com/kubernetes/kubernetes/pull/64326.
We can't skip it now as we support v1.10 clusters still.

luxas commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -141,9 +139,11 @@ func createStaticPodFiles(manifestDir string, cfg *kubeadmapi.MasterConfiguratio
+		"advertise-address": cfg.API.AdvertiseAddress,
+		"insecure-port":     "0",
+		// https://github.com/kubernetes/kubernetes/pull/58123
+		"enable-admission-plugin":         "NodeRestriction",
```
These are actually called enable/disable-admission-plugin**s**

luxas reviewed (commented):
Please fix these comments. After that this LGTM

neolit123 commented:
updated.
- fixed typo - missing `s` in argument keys
- add TODO comment above `PersistentVolumeLabel` usage

added the same TODO note here and also made this PR not close the issue but rather update it.
https://github.com/kubernetes/kubeadm/issues/840

luxas reviewed (approved):
/lgtm

luxas commented:
/retest

neolit123 commented:
/test pull-kubernetes-e2e-gce-100-performance
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-e2e-gce-device-plugin-gpu

Arnavion commented:
Using minikube to create a 1.11.0 cluster puts this in `/etc/kubernetes/manifests/kube-apiserver.yaml` :

```yaml
spec:
  containers:
  - command:
    - kube-apiserver
    - --admission-control=Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota
    <snip>
    - --disable-admission-plugins=PersistentVolumeLabel
    - --enable-admission-plugins=NodeRestriction
```

so apiserver fails to start because `admission-control and enable-admission-plugins/disable-admission-plugins flags are mutually exclusive`

```
$ minikube version

minikube version: v0.28.0
```

and inside the cluster VM:

```
$ kubeadm version 

kubeadm version: &version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:14:41Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
```

The enable/disable parameters are presumably coming because of this PR, so is there some place left that's still creating the default admission control parameter list?

neolit123 commented:
the `--admission-control` part should not be there. it's unclear to me what's adding it.

Arnavion commented:
Ah, it's coming from the MasterConfiguration in the config file passed to kubeadm. So this is probably minikube creating the file incorrectly. Will follow up there.

neolit123 commented:
@Arnavion 
yes, these come from Minikube:
https://github.com/kubernetes/minikube/blob/master/pkg/util/constants.go#L35
https://github.com/kubernetes/minikube/blob/master/pkg/minikube/bootstrapper/kubeadm/versions.go#L205

this usage is out of date with k8s 1.11.

Arnavion commented:
Yeah, just found it. I'll file a bug there.
